File: core-android/src/main/java/com/uber/sdk/android/core/utils/CustomTabsHelper.java
// Copyright 2015 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.uber.sdk.android.core.utils;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.customtabs.CustomTabsClient;
import android.support.customtabs.CustomTabsIntent;
import android.support.customtabs.CustomTabsServiceConnection;
import android.text.TextUtils;
import android.util.Log;

import com.uber.sdk.android.core.UberSdk;

import java.util.ArrayList;
import java.util.List;

/**
 * Helper class for Custom Tabs.
 */
public class CustomTabsHelper {
    private static final String TAG = "CustomTabsHelper";
    static final String STABLE_PACKAGE = "com.android.chrome";
    static final String BETA_PACKAGE = "com.chrome.beta";
    static final String DEV_PACKAGE = "com.chrome.dev";
    static final String LOCAL_PACKAGE = "com.google.android.apps.chrome";
    private static final String EXTRA_CUSTOM_TABS_KEEP_ALIVE =
            "android.support.customtabs.extra.KEEP_ALIVE";
    private static final String ACTION_CUSTOM_TABS_CONNECTION =
            "android.support.customtabs.action.CustomTabsService";

    private static String packageNameToUse;

    private CustomTabsServiceConnection connection;

    public CustomTabsHelper() {}

    /**
     * Opens the URL on a Custom Tab if possible. Otherwise fallsback to opening it on a WebView.
     *
     * @param context The host context.
     * @param customTabsIntent a CustomTabsIntent to be used if Custom Tabs is available.
     * @param uri the Uri to be opened.
     * @param fallback a CustomTabFallback to be used if Custom Tabs is not available.
     */
    public void openCustomTab(
            final Context context,
            final CustomTabsIntent customTabsIntent,
            final Uri uri,
            CustomTabFallback fallback) {
        final String packageName = getPackageNameToUse(context);

        if (packageName != null) {
            connection = new CustomTabsServiceConnection() {
                @Override
                public void onCustomTabsServiceConnected(ComponentName componentName, CustomTabsClient client) {
                    client.warmup(0L); // This prevents backgrounding after redirection

                    customTabsIntent.intent.setPackage(packageName);
                    customTabsIntent.intent.setData(uri);
                    customTabsIntent.launchUrl(context, uri);
                }

                @Override
                public void onServiceDisconnected(ComponentName name) {
                }
            };
            CustomTabsClient.bindCustomTabsService(context, packageName, connection);
        } else if (fallback != null) {
            fallback.openUri(context, uri);
        } else {
            Log.e(UberSdk.UBER_SDK_LOG_TAG, "Use of openCustomTab without Customtab support or a fallback set");
        }
    }

    /**
     * Called to clean up the CustomTab when the parentActivity is destroyed.
     */
    public void onDestroy(Activity parentActivity) {
        if (connection != null) {
            parentActivity.unbindService(connection);
            connection = null;
        }
    }

    /**
     * Goes through all apps that handle VIEW intents and have a warmup service. Picks
     * the one chosen by the user if there is one, otherwise makes a best effort to return a
     * valid package name.
     *
     * This is <strong>not</strong> threadsafe.
     *
     * @param context {@link Context} to use for accessing {@link PackageManager}.
     * @return The package name recommended to use for connecting to custom tabs related components.
     */
    @Nullable
    public String getPackageNameToUse(Context context) {
        if (packageNameToUse != null) return packageNameToUse;

        PackageManager pm = context.getPackageManager();
        // Get default VIEW intent handler.
        Intent activityIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("http://www.example.com"));
        ResolveInfo defaultViewHandlerInfo = pm.resolveActivity(activityIntent, 0);
        String defaultViewHandlerPackageName = null;
        if (defaultViewHandlerInfo != null) {
            defaultViewHandlerPackageName = defaultViewHandlerInfo.activityInfo.packageName;
        }

        // Get all apps that can handle VIEW intents.
        List<ResolveInfo> resolvedActivityList = pm.queryIntentActivities(activityIntent, 0);
        List<String> packagesSupportingCustomTabs = new ArrayList<>();
        for (ResolveInfo info : resolvedActivityList) {
            Intent serviceIntent = new Intent();
            serviceIntent.setAction(ACTION_CUSTOM_TABS_CONNECTION);
            serviceIntent.setPackage(info.activityInfo.packageName);
            if (pm.resolveService(serviceIntent, 0) != null) {
                packagesSupportingCustomTabs.add(info.activityInfo.packageName);
            }
        }

        // Now packagesSupportingCustomTabs contains all apps that can handle both VIEW intents
        // and service calls.
        if (packagesSupportingCustomTabs.isEmpty()) {
            packageNameToUse = null;
        } else if (packagesSupportingCustomTabs.size() == 1) {
            packageNameToUse = packagesSupportingCustomTabs.get(0);
        } else if (!TextUtils.isEmpty(defaultViewHandlerPackageName)
                && !hasSpecializedHandlerIntents(context, activityIntent)
                && packagesSupportingCustomTabs.contains(defaultViewHandlerPackageName)) {
            packageNameToUse = defaultViewHandlerPackageName;
        } else if (packagesSupportingCustomTabs.contains(STABLE_PACKAGE)) {
            packageNameToUse = STABLE_PACKAGE;
        } else if (packagesSupportingCustomTabs.contains(BETA_PACKAGE)) {
            packageNameToUse = BETA_PACKAGE;
        } else if (packagesSupportingCustomTabs.contains(DEV_PACKAGE)) {
            packageNameToUse = DEV_PACKAGE;
        } else if (packagesSupportingCustomTabs.contains(LOCAL_PACKAGE)) {
            packageNameToUse = LOCAL_PACKAGE;
        }
        return packageNameToUse;
    }

    /**
     * Used to check whether there is a specialized handler for a given intent.
     * @param intent The intent to check with.
     * @return Whether there is a specialized handler for the given intent.
     */
    private boolean hasSpecializedHandlerIntents(Context context, Intent intent) {
        try {
            PackageManager pm = context.getPackageManager();
            List<ResolveInfo> handlers = pm.queryIntentActivities(
                    intent,
                    PackageManager.GET_RESOLVED_FILTER);
            if (handlers == null || handlers.size() == 0) {
                return false;
            }
            for (ResolveInfo resolveInfo : handlers) {
                IntentFilter filter = resolveInfo.filter;
                if (filter == null) continue;
                if (filter.countDataAuthorities() == 0 || filter.countDataPaths() == 0) continue;
                if (resolveInfo.activityInfo == null) continue;
                return true;
            }
        } catch (RuntimeException e) {
            Log.e(TAG, "Runtime exception while getting specialized handlers");
        }
        return false;
    }

    /**
     * @return All possible chrome package names that provide custom tabs feature.
     */
    public String[] getPackages() {
        return new String[]{"", STABLE_PACKAGE, BETA_PACKAGE, DEV_PACKAGE, LOCAL_PACKAGE};
    }

    /**
     * Fallback that uses browser
     */
    public static class BrowserFallback implements CustomTabFallback {
        @Override
        public void openUri(@NonNull Context context, Uri uri) {
            Intent intent = new Intent(Intent.ACTION_VIEW, uri);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
        }
    }

    /**
     * To be used as a fallback to open the Uri when Custom Tabs is not available.
     */
    public interface CustomTabFallback {
        /**
         *
         * @param context The Context that wants to open the Uri.
         * @param uri The uri to be opened by the fallback.
         */
        void openUri(@NonNull Context context, Uri uri);
    }
}

File: core/src/main/kotlin/com/uber/sdk2/core/utils/CustomTabsHelper.kt
/**
 * Copyright (c) 2024 Uber Technologies, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.uber.sdk2.core.utils

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.net.Uri
import android.text.TextUtils
import android.util.Log
import androidx.browser.customtabs.CustomTabsClient
import androidx.browser.customtabs.CustomTabsIntent
import androidx.browser.customtabs.CustomTabsServiceConnection

/** Helper class for Custom Tabs. */
object CustomTabsHelper {
  private var connection: CustomTabsServiceConnection? = null

  /**
   * Opens the URL on a Custom Tab if possible. Otherwise fallsback to opening it on a WebView.
   *
   * @param context The host context.
   * @param customTabsIntent a CustomTabsIntent to be used if Custom Tabs is available.
   * @param uri the Uri to be opened.
   * @param fallback a CustomTabFallback to be used if Custom Tabs is not available.
   */
  fun openCustomTab(
    context: Context,
    customTabsIntent: CustomTabsIntent,
    uri: Uri,
    fallback: CustomTabFallback?,
  ) {
    val packageName = getPackageNameToUse(context)
    if (packageName != null) {
      val connection =
        object : CustomTabsServiceConnection() {
          override fun onCustomTabsServiceConnected(
            componentName: ComponentName,
            client: CustomTabsClient,
          ) {
            client.warmup(0L) // This prevents backgrounding after redirection
            customTabsIntent.intent.setPackage(packageName)
            customTabsIntent.intent.setData(uri)
            customTabsIntent.launchUrl(context, uri)
          }

          override fun onServiceDisconnected(name: ComponentName?) {}
        }
      CustomTabsClient.bindCustomTabsService(context, packageName, connection)
      this.connection = connection
    } else
      fallback?.openUri(context, uri)
        ?: Log.e(
          UBER_AUTH_LOG_TAG,
          "Use of openCustomTab without Customtab support or a fallback set",
        )
  }

  /** Called to clean up the CustomTab when the parentActivity is destroyed. */
  fun onDestroy(parentActivity: Activity) {
    connection?.let { parentActivity.unbindService(it) }
    connection = null
  }

  /**
   * Goes through all apps that handle VIEW intents and have a warmup service. Picks the one chosen
   * by the user if there is one, otherwise makes a best effort to return a valid package name.
   *
   * This is **not** threadsafe.
   *
   * @param context [Context] to use for accessing [PackageManager].
   * @return The package name recommended to use for connecting to custom tabs related components.
   */
  private fun getPackageNameToUse(context: Context): String? {
    if (packageNameToUse != null) return packageNameToUse
    val pm: PackageManager = context.packageManager
    // Get default VIEW intent handler.
    val activityIntent = Intent(Intent.ACTION_VIEW, Uri.parse("http://www.example.com"))
    val defaultViewHandlerInfo: ResolveInfo? = pm.resolveActivity(activityIntent, 0)
    var defaultViewHandlerPackageName: String? = null
    if (defaultViewHandlerInfo != null) {
      defaultViewHandlerPackageName = defaultViewHandlerInfo.activityInfo.packageName
    }

    // Get all apps that can handle VIEW intents.
    val resolvedActivityList: List<ResolveInfo> = pm.queryIntentActivities(activityIntent, 0)
    val packagesSupportingCustomTabs: MutableList<String?> = ArrayList()
    for (info in resolvedActivityList) {
      val serviceIntent = Intent()
      serviceIntent.setAction(ACTION_CUSTOM_TABS_CONNECTION)
      serviceIntent.setPackage(info.activityInfo.packageName)
      if (pm.resolveService(serviceIntent, 0) != null) {
        packagesSupportingCustomTabs.add(info.activityInfo.packageName)
      }
    }

    // Now packagesSupportingCustomTabs contains all apps that can handle both VIEW intents
    // and service calls.
    packageNameToUse =
      when {
        packagesSupportingCustomTabs.isEmpty() -> null
        packagesSupportingCustomTabs.size == 1 -> packagesSupportingCustomTabs[0]
        !TextUtils.isEmpty(defaultViewHandlerPackageName) &&
          !hasSpecializedHandlerIntents(context, activityIntent) &&
          packagesSupportingCustomTabs.contains(defaultViewHandlerPackageName) ->
          defaultViewHandlerPackageName
        packagesSupportingCustomTabs.contains(STABLE_PACKAGE) -> STABLE_PACKAGE
        packagesSupportingCustomTabs.contains(BETA_PACKAGE) -> BETA_PACKAGE
        else -> packagesSupportingCustomTabs[0]
      }
    return packageNameToUse
  }

  /**
   * Used to check whether there is a specialized handler for a given intent.
   *
   * @param intent The intent to check with.
   * @return Whether there is a specialized handler for the given intent.
   */
  private fun hasSpecializedHandlerIntents(context: Context, intent: Intent): Boolean {
    try {
      val pm: PackageManager = context.packageManager
      val handlers: List<ResolveInfo> =
        pm.queryIntentActivities(intent, PackageManager.GET_RESOLVED_FILTER)
      if (handlers.isEmpty()) {
        return false
      }
      handlers.forEach { resolveInfo ->
        resolveInfo.filter?.let { filter ->
          if (
            filter.countDataAuthorities() != 0 &&
              filter.countDataPaths() != 0 &&
              resolveInfo.activityInfo != null
          ) {
            return true // A suitable handler is found, return true immediately
          }
        }
      }
    } catch (e: RuntimeException) {
      Log.e(TAG, "Runtime exception while getting specialized handlers")
    }
    return false
  }

  /** Fallback that uses browser */
  class BrowserFallback : CustomTabFallback {
    override fun openUri(context: Context, uri: Uri?) {
      val intent = Intent(Intent.ACTION_VIEW, uri)
      intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
      context.startActivity(intent)
    }
  }

  /** To be used as a fallback to open the Uri when Custom Tabs is not available. */
  interface CustomTabFallback {
    /**
     * @param context The Context that wants to open the Uri.
     * @param uri The uri to be opened by the fallback.
     */
    fun openUri(context: Context, uri: Uri?)
  }

  private const val TAG = "CustomTabsHelper"
  private const val STABLE_PACKAGE = "com.android.chrome"
  private const val BETA_PACKAGE = "com.chrome.beta"
  private const val ACTION_CUSTOM_TABS_CONNECTION =
    "android.support.customtabs.action.CustomTabsService"
  private var packageNameToUse: String? = null

  private const val UBER_AUTH_LOG_TAG = "UberAuth"
}
